core/src/main/java/apoc/meta/Meta.java

```diff
@@ -909,39 +909,59 @@ private void filterNonExistingRelationships(Map<Pattern, Relationship> vRels, Me
+    private boolean relationshipExists(Pattern p, Relationship relationship, SampleMetaConfig metaConfig) {
+        if (relationship == null) return false;
+        double degreeFrom = (double)(long)relationship.getProperty("out")  / (long) relationship.getStartNode().getProperty("count");
+        double degreeTo = (double)(long)relationship.getProperty("in")  / (long) relationship.getEndNode().getProperty("count");
 
+        Map<String, Long> countStore = getLabelCountStore(tx, kernelTx);
         if (degreeFrom < degreeTo) {
-            if (relationshipExists(p.labelFrom(), p.labelTo(), p.relationshipType(), Direction.OUTGOING, metaConfig)) return true;
+            return relationshipExists(tx, countStore, p.labelFrom(), p.labelTo(), p.relationshipType(), Direction.OUTGOING, metaConfig);
         } else {
-            if (relationshipExists(p.labelTo(), p.labelFrom(), p.relationshipType(), Direction.INCOMING, metaConfig)) return true;
+            return relationshipExists(tx, countStore, p.labelTo(), p.labelFrom(), p.relationshipType(), Direction.INCOMING, metaConfig);
         }
-        return false;
     }
 
-    private boolean relationshipExists(Label labelFromLabel, Label labelToLabel, RelationshipType relationshipType, Direction direction, MetaConfig metaConfig) {
-        Map<String, Long> countStore = getLabelCountStore();
+    /**
+     * relationshipExists uses sampling to check if the relationships added in previous steps exist.
+     * The sample count is the skip count; e.g. if set to 1000 this means every 1000th node will be checked.
+     * A sample count higher than nodeLabel count is equivalent to not checking them.
+     * Note; Each node is still fetched, but the relationships on that node will not be checked
+     * if skipped, which should make it faster.
+     */
+    static boolean relationshipExists(
+            Transaction tx,
+            Map<String, Long>  countStore,
+            Label labelFromLabel,
+            Label labelToLabel,
+            RelationshipType relationshipType,
+            Direction direction,
+            SampleMetaConfig metaConfig
+    ) {
         try (ResourceIterator<Node> nodes = tx.findNodes(labelFromLabel)) {
-            long count = 1L;
-            String labelName = labelFromLabel.name();
-            long labelCount = countStore.get(labelName);
-            long sample = getSampleForLabelCount(labelCount, metaConfig.getSample());
+            long count = 0L;
+            long labelCount = countStore.get(labelFromLabel.name());
+            // A sample size below 0 means we should check every node.
+            long skipCount = metaConfig.getSample() > 0 ? metaConfig.getSample() : 1;
+            // Sample size greater than label count is too large, skip
+            // Returning true will keep the relationship, otherwise it may remove all relationships.
+            if (skipCount > labelCount) return true;
```

Comment:
I still find it a bit weird that if we have a labelCount = 1000 and a skipCount = 1000 we will check 1 node only and probably filter out most relationships, while if we have skipCount = 1001 it would be a whole other behaviour where we answer true for everything.

An alternative could be to do the count++ last in the loop, then we would check exactly one node (the first one when count is 0) for skipCount >= labelCount. What do you think?